### PR TITLE
added ComboBox component to the library

### PR DESCRIPTION
**Pull request by sahandfarahani**
ComboBoxgif

### CODE REVIEW

File: src/Inputs/ComboBox.tsx [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    const handleChange=(event:any)=>{
+        myWindow.inputValue=event.target.value;
+        setExpanded(true);
+    }
+
+    return (
+        <LabelLayout {...props}>
+            <Container>
+
+                    <InputFragment value={myWindow.inputValue}
```
Comment (ralph-dev) [inline src/Inputs/ComboBox.tsx:146]: The indentation looks off?

File: src/Inputs/ComboBox.tsx [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+                                } 
+                    onClick={e=>
+                        handleChange(e)
+                        } 
+                        
+                     />
+
+                {mount && (
+                    <SelectList
+                        limit={Math.min(3, limit)}
```
Comment (ralph-dev) [inline src/Inputs/ComboBox.tsx:161]: 3 is a magic number, use a constant NUM_VISIBLE_SELECTIONS = 3;

File: src/Inputs/ComboBox.tsx (hunk 1 above)
Comment (sahandfarahani) [reply src/Inputs/ComboBox.tsx:146]: sorry
fixed that

File: src/Inputs/ComboBox.tsx (hunk 2 above)
Comment (sahandfarahani) [reply src/Inputs/ComboBox.tsx]: done

File: src/Inputs/ComboBox.tsx [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+const _Select: React.FC<SelectProps> = ({
+    disabled,
+    value,
+    children,
+    limit = 4,
+    placeholder = '',
+    onChange = (): void => {},
+    theme,
+    name,
+    NUM_VISIBLE_SELECTIONS,
```
Comment (ralph-dev) [inline src/Inputs/ComboBox.tsx:45]: This should be a variable not a constant?

Meaning
, numVisible = NUM_VISIBLE_SELECTIONS

and at line 21, const NUM_VISIBLE_SELECTIONS = 3;

Comment (sahandfarahani) [reply src/Inputs/ComboBox.tsx]: I think I did it as mentioned now

File: src/Inputs/ComboBox.tsx [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    position,
+    transition,
+    clickable,
+    styledCondition,
+    darken,
+} from '@Utils/Mixins';
+import { useTransition } from '@Utils/Hooks';
+import { LabelLayout, LabelLayoutProps,InputFragment } from '@Layouts';
+ 
+const NUM_VISIBLE_SELECTIONS=3;
```
Comment (ralph-dev) [inline src/Inputs/ComboBox.tsx:22]:
```suggestion
const NUM_VISIBLE_SELECTIONS = 3;

```

File: src/Inputs/ComboBox.tsx [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+const _Select: React.FC<SelectProps> = ({
+    disabled,
+    value,
+    children,
+    limit = 4,
+    placeholder = '',
+    onChange = (): void => {},
+    theme,
+    name,
+    numVisible=NUM_VISIBLE_SELECTIONS,
```
Comment (ralph-dev) [inline src/Inputs/ComboBox.tsx:47]:
```suggestion
    numVisible = NUM_VISIBLE_SELECTIONS,

```

File: src/Inputs/ComboBox.tsx [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+    ...props
+}): React.ReactElement => {
+    const [expanded, setExpanded] = useState(false);
+    const options = Children.toArray(children);
```
Comment (ralph-dev) [inline src/Inputs/ComboBox.tsx:51]: This will run every render, should this be in a useEffect?

File: src/Inputs/ComboBox.tsx [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+            });
+            myWindow.inputValue=currentTarget.firstChild.nodeValue;
+        },
+        [name],
+    );
+
+    useEffect((): void | (() => void | undefined) => {
+        if (expanded) {
+            const listener = (): void => {
+                setExpanded(false);
```
Comment (ralph-dev) [inline src/Inputs/ComboBox.tsx:122]: Where is the isMounted control?

File: src/Inputs/ComboBox.tsx [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+            };
+            const timer = window.setTimeout((): void => {
```
Comment (ralph-dev) [inline src/Inputs/ComboBox.tsx:124]: Also, let's name these timers more properly

Comment (sahandfarahani) [reply src/Inputs/ComboBox.tsx]: all done

File: src/Inputs/ComboBox.tsx [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+                window.clearTimeout(timerforkeydown);
+                window.clearTimeout(timerforclick);
+                window.removeEventListener('keydown', listener);
+                window.removeEventListener('click', listener);
+            };
+        }
+        return undefined;
+    }, [expanded]);
+
+    const handleChange=(event:any)=>{
```
Comment (nikitasheremet) [inline src/Inputs/ComboBox.tsx:140]: (event: SyntheticEvent), import Synthetic Event from react

File: src/Inputs/ComboBox.tsx [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+                    value: currentTarget.getAttribute('value'),
+                    name,
+                },
+            });
+            myWindow.inputValue=currentTarget.firstChild.nodeValue;
+        },
+        [name],
+    );
+
+    useEffect((): void | (() => void | undefined) => {
```
Comment (nikitasheremet) [inline src/Inputs/ComboBox.tsx:113]: You do not need to add any extra typings for useEffect, as far as I know

File: src/Inputs/ComboBox.tsx [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+            }, 10);
+            return (): void => {
+                isMounted=false;
+                window.clearTimeout(timerforkeydown);
+                window.clearTimeout(timerforclick);
+                window.removeEventListener('keydown', listener);
+                window.removeEventListener('click', listener);
+            };
+        }
+        return undefined;
```
Comment (nikitasheremet) [inline src/Inputs/ComboBox.tsx:137]: The correct return would be return = () => isMounted = false
A useEffect returns a clean up function ( () => {} )

Comment (sahandfarahani) [reply src/Inputs/ComboBox.tsx:137]: so I should remove isMounted=false from line130 , correct?

File: src/Inputs/ComboBox.tsx [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+    disabled?: boolean;
+    placeholder?: string;
+    value?: string | number;
+    theme: DefaultTheme;
+    onChange?: Function;
+    limit?: number;
+    numVisible:number;
+}
+
+const _Select: React.FC<SelectProps> = ({
```
Comment (fpunny) [inline src/Inputs/ComboBox.tsx:41]: Why does combobox copy code from select and not use select?

File: src/Inputs/ComboBox.tsx [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+    transition,
+    clickable,
+    darken,
+} from '@Utils/Mixins';
+import { useTransition } from '@Utils/Hooks';
+import { LabelLayout, LabelLayoutProps,InputFragment } from '@Layouts';
+
+const NUM_VISIBLE_SELECTIONS = 3;
+
+var options:any = [];
```
Comment (fpunny) [inline src/Inputs/ComboBox.tsx:23]: No >:c bad

File: src/Inputs/ComboBox.tsx (hunk 6 above)
Comment (fpunny) [reply src/Inputs/ComboBox.tsx]: useMemo

File: src/Inputs/ComboBox.tsx (hunk 9 above)
Comment (fpunny) [reply src/Inputs/ComboBox.tsx]: uhhhh React.ClickEvent or something like that is better if you can

File: src/Inputs/ComboBox.tsx [hunk 14]
```diff
@@ -117,19 +111,25 @@ const _Select: React.FC<SelectProps> = ({
     );
 
     useEffect((): void | (() => void | undefined) => {
+        isMounted=true;
```
Comment (ralph-dev) [inline src/Inputs/ComboBox.tsx:115]: let isMounted

File: src/Inputs/ComboBox.tsx [hunk 15]
```diff
@@ -13,13 +12,16 @@ import {
...
 import { useTransition } from '@Utils/Hooks';
 import { LabelLayout, LabelLayoutProps,InputFragment } from '@Layouts';
- 
-const NUM_VISIBLE_SELECTIONS=3;
+
+const NUM_VISIBLE_SELECTIONS = 3;
+
+var options:any = [];
+
+var isMounted=false;
```
Comment (ralph-dev) [inline src/Inputs/ComboBox.tsx:25]: declare inside the useEffect

File: src/Inputs/ComboBox.tsx (hunk 11 above)
Comment (fpunny) [reply src/Inputs/ComboBox.tsx]: @nikitasheremet mans learning owo

Comment (fpunny) [reply src/Inputs/ComboBox.tsx]: @sahandfarahani ur that is correct, but ur implementation of isMounted needs work and should also be set to false on line 137

File: src/Inputs/ComboBox.tsx (hunk 10 above)
Comment (fpunny) [reply src/Inputs/ComboBox.tsx]: unfortunately u do since we have function return being required

File: src/Inputs/ComboBox.tsx (hunk 9 above)
Comment (sahandfarahani) [reply src/Inputs/ComboBox.tsx]: > @fpunny well it gets executed whether Click or Keydown happens so maybe a more general type is better

File: src/Inputs/ComboBox.tsx (hunk 13 above)
Comment (sahandfarahani) [reply src/Inputs/ComboBox.tsx:23]: :)))
everything from 23 to 29?

File: src/Inputs/ComboBox.tsx (hunk 12 above)
Comment (sahandfarahani) [reply src/Inputs/ComboBox.tsx:41]: its mostly select, I'm going to delete the unnecessary props that were useless in the combobox

File: src/Inputs/ComboBox.tsx (hunk 13 above)
Comment (sahandfarahani) [reply src/Inputs/ComboBox.tsx:23]: I'm using options array in a useMemo and changing window type from any to Window & typeof globalThis

Comment (sahandfarahani) [reply src/Inputs/ComboBox.tsx:23]: I'm using options in a useMemo hook now

File: src/Inputs/ComboBox.tsx (hunk 12 above)
Comment (fpunny) [reply src/Inputs/ComboBox.tsx:36]: No but why are we copying the select component code and not using the select component directly

**Pull request merged**